src/builder.rs

```diff
@@ -0,0 +1,304 @@
+    /// This can be used to allocate a buffer for the `build_into` method.
+    pub fn target_length(&self) -> usize {
+        /* 12 is the length of the basic header */
+        let mut length = 12usize;
+        length += self.csrc_count as usize * 4;
+        length += if let Some((_, ext)) = self.extension { ext.len() + 4 } else { 0 };
+        length += if let Some(payload) = self.payload { payload.len() } else { 0 };
+        if self.padded && (length & 0x3) != 0 {
+            length += (!(length & 0x3) + 1) & 0x3;
+        }
+        length
+    }
+
+    /// Build the packet into the target buffer but ignore all validity checks.
+    pub fn build_into_unchecked(&mut self, target: &mut [u8]) -> usize {
+        let first_byte = &mut target[0];
+        *first_byte = 2 << 6; /* The RTP packet version */
+        if self.extension.is_some() {
+            *first_byte |= 1 << 4; /* set the extension flag */
+        }
+        *first_byte |= self.csrc_count;
+
+        target[1] = self.payload_type;
+        if self.marked {
+            target[1] |= 0x80;
+        }
+
+        target[ 2] = (self.sequence.0 >>    8) as u8;
+        target[ 3] = (self.sequence.0 &  0xFF) as u8;
+
+        target[ 4] = (self.timestamp >> 24) as u8;
+        target[ 5] = (self.timestamp >> 16) as u8;
+        target[ 6] = (self.timestamp >>  8) as u8;
+        target[ 7] = (self.timestamp      ) as u8;
+
+        target[ 8] = (self.ssrc >> 24) as u8;
+        target[ 9] = (self.ssrc >> 16) as u8;
+        target[10] = (self.ssrc >>  8) as u8;
+        target[11] = (self.ssrc      ) as u8;
+
+        let mut write_index = 12usize;
+        for index in 0..self.csrc_count as usize {
+            let csrc = self.csrcs[index];
+            target[write_index + 0] = (csrc >> 24) as u8;
+            target[write_index + 1] = (csrc >> 16) as u8;
+            target[write_index + 2] = (csrc >>  8) as u8;
+            target[write_index + 3] = (csrc      ) as u8;
+
+            write_index = write_index + 4;
+        }
+
+        if let Some((id, payload)) = self.extension {
+            target[write_index + 0] = (id >>    8) as u8;
+            target[write_index + 1] = (id &  0xFF) as u8;
+
+            let len = payload.len() / 4;
+            target[write_index + 2] = (len >>    8) as u8;
+            target[write_index + 3] = (len &  0xFF) as u8;
+
+            write_index = write_index + 4;
+
+            /* the target buffer has been ensured to hold that many bytes */
+            target[write_index..(write_index + payload.len())].copy_from_slice(payload);
+            write_index += payload.len();
+        }
+
+        if let Some(payload) = self.payload {
+            /* the target buffer has been ensured to hold that many bytes */
+            target[write_index..(write_index + payload.len())].copy_from_slice(payload);
+            write_index += payload.len();
+        }
+
+        if self.padded && (write_index & 0x3) != 0 {
```

Comment:
A bit late I know, but I can't spot a place in rfc3550 requiring that when padding is present, it will only pad the packet length to a multiple of 4 bytes, which I think is what the builder does here.

Looking back, I believe my assumption when writing the parser was that an arbitrary amount of padding from 1-255 bytes is allowed.

Do you think I misunderstood, or do you agree that for a fully general API caller need to specify an amount of padding here?  Maybe padding property should be `Option<u8>` rather than `bool` for both reader and builder.

This only comes up because I'm attempting to wire the builder into the fuzz-test, asserting that we get back the bytes we started with. This is revealing anywhere that reader and builder make different assumptions!  :sweat_smile: